src/Adapter/Order/CommandHandler/AddProductToOrderHandler.php

```diff
@@ -0,0 +1,506 @@
+            $command,
+            $order,
+            $cart,
+            $product,
+            $combination
+        );
+
+        $this->addProductToCart($cart, $product, $combination, $command->getProductQuantity());
+
+        $invoice = $this->createNewOrEditExistingInvoice(
+            $command,
+            $order,
+            $cart
+        );
+
+        $totalMethod = $command->getOrderInvoiceId() ? Cart::BOTH_WITHOUT_SHIPPING : Cart::BOTH;
+
+        // Create Order detail information
+        $orderDetail = new OrderDetail();
+        $orderDetail->createList(
+            $order,
+            $cart,
+            $order->getCurrentOrderState(),
+            $cart->getProducts(),
+            $command->getOrderInvoiceId() ?: 0
+        );
+
+        // update totals amount of order
+        $order->total_products += (float) $cart->getOrderTotal(false, Cart::ONLY_PRODUCTS);
+        $order->total_products_wt += (float) $cart->getOrderTotal(true, Cart::ONLY_PRODUCTS);
+
+        $order->total_paid += Tools::ps_round((float) $cart->getOrderTotal(true, $totalMethod), 2);
+        $order->total_paid_tax_excl += Tools::ps_round((float) $cart->getOrderTotal(false, $totalMethod), 2);
+        $order->total_paid_tax_incl += Tools::ps_round((float) $cart->getOrderTotal(true, $totalMethod), 2);
+
+        if (null !== $invoice && Validate::isLoadedObject($invoice)) {
+            $order->total_shipping = $invoice->total_shipping_tax_incl;
+            $order->total_shipping_tax_incl = $invoice->total_shipping_tax_incl;
+            $order->total_shipping_tax_excl = $invoice->total_shipping_tax_excl;
+        }
+
+        // discount
+        $order->total_discounts += (float) abs($cart->getOrderTotal(true, Cart::ONLY_DISCOUNTS));
+        $order->total_discounts_tax_excl += (float) abs($cart->getOrderTotal(false, Cart::ONLY_DISCOUNTS));
+        $order->total_discounts_tax_incl += (float) abs($cart->getOrderTotal(true, Cart::ONLY_DISCOUNTS));
+
+        // Save changes of order
+        $order->update();
+
+        StockAvailable::synchronize($product->id);
+
+        // Update weight SUM
+        $orderCarrier = new OrderCarrier((int) $order->getIdOrderCarrier());
+        if (Validate::isLoadedObject($orderCarrier)) {
+            $orderCarrier->weight = (float) $order->getTotalWeight();
+            if ($orderCarrier->update()) {
+                $order->weight = sprintf('%.3f ' . Configuration::get('PS_WEIGHT_UNIT'), $orderCarrier->weight);
+            }
+        }
+
+        // Update Tax lines
+        $orderDetail->updateTaxAmount($order);
+
+        // Delete specific price if exists
+        if (null !== $specificPrice) {
+            $specificPrice->delete();
+        }
+
+        $order = $order->refreshShippingCost();
+
+        Hook::exec('actionOrderEdited', ['order' => $order]);
+
+        $oldCartRules = Context::getContext()->cart->getCartRules();
+        $newCartRules = Context::getContext()->cart->getCartRules();
```

Comment:
😮 what's the difference between `$oldCartRules` and `$newCartRules` ?

Reply:
I've moved `$oldCartRules` to the top since they had to be retrieved from an old cart (because new cart is created somewhere in between).

Reply:
But now there's nothing in between 🤔

Reply:
There will be after I push changes.